src/Stratis.Features.FederatedPeg/InputConsolidation/InputConsolidator.cs

```diff
@@ -0,0 +1,301 @@
+            }
+
+            // Loop exits when we get a set of 50 that had a high enough amount, or when we run out of UTXOs aka a round less than 50
+
+            return consolidationTransactions;
+        }
+
+        /// <summary>
+        /// Build a consolidating transaction.
+        /// </summary>
+        public Transaction BuildConsolidatingTransaction(List<UnspentOutputReference> selectedInputs)
+        {
+            try
+            {
+                string walletPassword = this.walletManager.Secret.WalletPassword;
+                bool sign = (walletPassword ?? "") != "";
+
+                var multiSigContext = new TransactionBuildContext(new List<Recipient>())
+                {
+                    MinConfirmations = WithdrawalTransactionBuilder.MinConfirmations,
+                    Shuffle = false,
+                    IgnoreVerify = true,
+                    WalletPassword = walletPassword,
+                    Sign = sign,
+                    TransactionFee = FederatedPegSettings.ConsolidationFee,
+                    SelectedInputs = selectedInputs.Select(u => u.ToOutPoint()).ToList(),
+                    AllowOtherInputs = false,
+                    IsConsolidatingTransaction = true,
+                    Time = (uint?) selectedInputs.First().Transaction.CreationTime.ToUnixTimeSeconds()
+                };
+
+                Transaction transaction = this.transactionHandler.BuildTransaction(multiSigContext);
+
+                this.logger.LogDebug("Consolidating transaction = {0}", transaction.ToString(this.network, RawFormat.BlockExplorer));
+
+                return transaction;
+            }
+            catch (Exception e)
+            {
+                this.logger.LogWarning("Exception when building consolidating transaction. Wallet state likely changed before calling: " + e);
+                return null;
+            }
+        }
+
+        /// <inheritdoc />
+        public void ProcessTransaction(Transaction transaction)
+        {
+            lock (this.lockObj)
+            {
+                // No work to do
+                if (this.ConsolidationTransactions == null)
+                    return;
+
+                // If we have a transaction in memory that is not FullySigned yet, set it to be as such when we receive a transaction from the mempool.
+                if (this.IsConsolidatingTransaction(transaction))
+                {
+                    ConsolidationTransaction inMemoryTransaction = this.GetInMemoryConsolidationTransaction(transaction);
+
+                    if (inMemoryTransaction != null && inMemoryTransaction.Status == ConsolidationTransactionStatus.Partial)
+                    {
+                        this.logger.LogDebug("Saw condensing transaction {0} in mempool, updating its status to FullySigned", transaction.GetHash());
+                        inMemoryTransaction.Status = ConsolidationTransactionStatus.FullySigned;
```

Comment:
Perhaps add a sanity check that it is in fact fully signed.

Reply:
Hmm, it takes quite a bit of time to verify large transactions like that, I'd rather avoid it here, they are only coming from the mempool at the moment.

Reply:
Ok